pkg/controller/pinned-image-set/pinned_image_set_controller.go

```diff
@@ -0,0 +1,702 @@
+		// check if not found error
+		isNotFound := apierrors.IsNotFound(err)
+		if err != nil && !isNotFound {
+			return err
+		}
+		// If we have seen this generation and the sync didn't fail, then skip
+		if !isNotFound && imageSet.Status.ObservedGeneration >= imageSet.Generation && mc.Annotations[ctrlcommon.GeneratedByControllerVersionAnnotationKey] != version.Hash {
+			// But we still need to compare the generated controller version because during an upgrade we need a new one
+			mcCtrlVersion := mc.Annotations[ctrlcommon.GeneratedByControllerVersionAnnotationKey]
+			if mcCtrlVersion == version.Hash {
+				return nil
+			}
+		}
+
+		pinnedImages := imageSet.Spec.PinnedImages
+		if len(pinnedImages) == 0 {
+			return fmt.Errorf("no pinned images for PinnedImageSet %s", imageSet.Name)
+		}
+
+		tomlConf := tomlConfigCRIOPinnedImages{}
+		tomlConf.Crio.Image.PinnedImages = pinnedImages
+		configFile, err := createTOMLgeneratedConfigFile(CRIODropInFilePathPinnedImages, tomlConf)
+		if err != nil {
+			return err
+		}
+
+		if isNotFound {
+			tempIgnCfg := ctrlcommon.NewIgnConfig()
+			mc, err = ctrlcommon.MachineConfigFromIgnConfig(pool.Name, managedKey, tempIgnCfg)
+			if err != nil {
+				return err
+			}
+		}
+
+		imageSetRuntimeConfigIgn := createNewIgnition(configFile)
+		rawCtrRuntimeConfigIgn, err := json.Marshal(imageSetRuntimeConfigIgn)
+		if err != nil {
+			return fmt.Errorf("error marshalling PinnedImageSet Ignition: %w", err)
+		}
+		mc.Spec.Config.Raw = rawCtrRuntimeConfigIgn
+
+		mc.SetAnnotations(map[string]string{
+			ctrlcommon.GeneratedByControllerVersionAnnotationKey: version.Hash,
+		})
+		oref := metav1.NewControllerRef(imageSet, controllerKind)
+		mc.SetOwnerReferences([]metav1.OwnerReference{*oref})
+
+		// Create or Update, on conflict retry
+		if err := retry.RetryOnConflict(ctrl.waitBackOff, func() error {
+			var err error
+			if isNotFound {
+				_, err = ctrl.mcfgClient.MachineconfigurationV1().MachineConfigs().Create(context.TODO(), mc, metav1.CreateOptions{})
+			} else {
+				_, err = ctrl.mcfgClient.MachineconfigurationV1().MachineConfigs().Update(context.TODO(), mc, metav1.UpdateOptions{})
+			}
+			return err
+		}); err != nil {
+			return fmt.Errorf("failed to create/update MachineConfig: %v", err)
+		}
+		// Add Finalizers to PinnedImageSet
+		if err := ctrl.addFinalizer(imageSet, mc); err != nil {
+			return fmt.Errorf("could not add finalizers to PinnedImageSet: %v", err)
+		}
+	}
+
+	return nil
+}
+
+func (ctrl *Controller) ensureNodeImagePrefetch(imageSet *mcfgv1.PinnedImageSet) error {
+	nodes, err := ctrl.getNodesForPinnedImageSet(imageSet)
+	if err != nil {
+		return err
+	}
+
+	if len(nodes) == 0 {
+		return fmt.Errorf("no nodes found for PinnedImageSet %s", imageSet.Name)
+	}
+
+	for _, node := range nodes {
+		if !ctrl.isPrefetchCompleteForNode(node, imageSet) {
+			// If prefetch is not complete fail fast and requeue the PinnedImageSet
```

Comment:
Could you help me understand this a bit. In the controller logic, you first `ensurePinnedImageSet` which deploys the machineconfig, then immediately after that sync this.

I assume expectation is that the image pulls should be taking awhile to complete, so is the expectation that the controller will be in an error state until the daemons are done?

Reply:
Good point, the expectation is that it should be in a Progressing state. As the error is expected. So we should adjust that.

Reply:
+1. We may want to show InProgress state to show that ImagePrefetch is progressing instead of an error.